pkg/auth/cred.go

```diff
@@ -17,19 +17,42 @@ limitations under the License.
 package auth
 
 import (
+	"context"
 	"fmt"
+	"time"
 
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
 	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
 	"k8s.io/klog/v2"
+	"knative.dev/pkg/logging"
 )
 
+type TokenWrapper struct {
+	cred azcore.TokenCredential
+}
+
+func NewTokenWrapper(cred azcore.TokenCredential) *TokenWrapper {
+	return &TokenWrapper{
+		cred: cred,
+	}
+}
+
+func (w *TokenWrapper) GetToken(ctx context.Context, options policy.TokenRequestOptions) (azcore.AccessToken, error) {
+	token, err := w.cred.GetToken(ctx, options)
+	if err != nil {
+		return azcore.AccessToken{}, err
+	}
+	logging.FromContext(ctx).Info("refreshing MDAL Token")
+	token.ExpiresOn = time.Now().Add(2 * time.Hour)
```

Comment:
My understanding is: the bug was `ExpiresOn` being set to be later than its real TTL, right?

Reply:
A more detailed description of the issue would be useful; would also help understand (and review) the mechanism of the fix.

Reply:
I think the bug is kind of related to the controller missed the refresh action due to unknown reasons (one possible cause could be clock skew). And for some reason, the controller cached the old token without really refreshing it (i guess this part of the [code](https://github.com/Azure/azure-sdk-for-go/blob/76e5495a4d1c7ed72e1c1dc973b126cc31cf88e4/sdk/internal/temporal/resource.go#L45-L114) could be potential culprit). The default refresh interval for the MI token from IMDS is around 23 hours, and expires on is set to 5 minutes before 23 hours. The idea of this fix is to reduce the token refresh interval to make it refresh often / earlier to address the potential clock skew.

Unfortunately, the original bug report doesn't come with some timestamp info about the original token expires on value and the actual system clock time etc, it's a little bit difficult to fully root cause here.

related discussion can be found from here: https://github.com/hashicorp/terraform-provider-azurerm/issues/20834

Reply:
Yeah i have a similar level of understanding its not perfect but thats the gist